File: Exercism Codes/JavaScript/run-length-encoding.js
//
// This is only a SKELETON file for the 'Run Length Encoding' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const encode = (string) => {
  let count = 1, output = "";

  for (let i = 1; i <= string.length; i++) {
    if (i < string.length && string[i] == string[i - 1]) {
      count++;
    }
    else {
      output += (count < 2 ? "" : count) + string[i - 1];
      count = 1;
    }
  }
  return output;
};

export const decode = (string) => {
  const matches = string.match(/(\d*[A-Z a-z])/g);

  if (matches === null) {
    return "";
  }
  
  let output = "";

  for (const match of matches) {
    const letter = match[match.length - 1];
    const number = match.substring(0, match.length - 1);
    output += letter.repeat(number.length === 0 ? 1 : Number(number));
  }

  return output;
};

File: Exercism Codes/Python/run_length_encoding.py
import re


def decode(string):
    matches = re.findall(r"(\d*[A-Z a-z])", string)
    output = ""

    for match in matches:
        letter = match[-1:]
        number = 1 if not match[:-1] else int(match[:-1])
        output += letter * number

    return output


def encode(string):
    count = 1
    output = ""
    for i in range(1, len(string) + 1):
        if i < len(string) and string[i] == string[i - 1]:
            count += 1
        else:
            output += ("" if count < 2 else str(count)) + string[i - 1]
            count = 1

    return output